### DISCUSSION TITLE
WARNING: lib not found: api-ms-win-core-path-l1-1-0.dll dependency of D:\Program Files\Python39\python39.dll

### DISCUSSION THREAD

**Discussion by aliencaocao** [pyinstaller/pyinstaller#6200] [PyInstaller] [answered]
`WARNING: lib not found: api-ms-win-core-path-l1-1-0.dll dependency of D:\Program Files\Python39\python39.dll`
When compiling using one file mode, i got this warning. Is this normal? So far I have not seen any side effects/errors in the compiled application, but I want to ensure stability.

Besides this, I also got
```
WARNING: Hidden import "pkg_resources.py2_warn" not found!
WARNING: Hidden import "pkg_resources.markers" not found!
```
Again no side effects seen so far.

Running pyinstaller 5.0dev0, python 3.9.7

**Comment by rokm** [comment, marked as answer]
Yeah, those are all harmless.

> `WARNING: lib not found: api-ms-win-core-path-l1-1-0.dll dependency of D:\Program Files\Python39\python39.dll`

The `api-ms-win-*` DLLs are typically not directly resolvable on Windows (unless you're using Anaconda python - in that case, Anaconda brings their own copy, so they are resolvable, and we collect them, even though perhaps we should not...).

I'll add a warning suppression rule for `api-ms-win-core.*` to prevent that warning from being shown.

> WARNING: Hidden import "pkg_resources.py2_warn" not found!
> WARNING: Hidden import "pkg_resources.markers" not found!

These are caused by hook adding those two modules to hiddenimports ([here](https://github.com/pyinstaller/pyinstaller/blob/708f5d90ff5fc91e1764ab082322ca7faf4d9c7b/PyInstaller/hooks/hook-pkg_resources.py#L20) and [here](https://github.com/pyinstaller/pyinstaller/blob/708f5d90ff5fc91e1764ab082322ca7faf4d9c7b/PyInstaller/hooks/hook-pkg_resources.py#L30)). These were probably applicable to earlier versions of `pkg_resources` but have since been removed, hence the warning.

**Reply by aliencaocao** [reply]
Ok thank you!

**Comment by liudonghua123** [comment]
I encountered the same problem when I packaged a pyqt5 app which use python 3.10 on my local win 11 pc and then send it to another person which use win 7.

I found the `api-ms-win-core-path-l1-1-0.dll` dependence was involved from python 3.9+. Old python depends `SHLWAPI.dll` instead of `api-ms-win-core-path-l1-1-0.dll`.

![image](https://user-images.githubusercontent.com/2276718/180375895-24f56289-759e-4e96-92e6-3b55bee33a7a.png)

**Reply by rokm** [reply]
Which is why [python.org download page](https://www.python.org/downloads/windows/) specifically tells you that python 3.9+ cannot be used on Windows 7: `"Note that Python 3.10.5 cannot be used on Windows 7 or earlier."`

**Reply by liudonghua123** [reply]
So, If I need to make an app which is compatible with windows 7+, I have only to use old python like 3.8?

**Reply by rokm** [reply]
Yes.

**Reply by rokm** [reply]
And if you want to be 100% certain that everything is collected, you should be building your app under Windows 7, too (e.g., in a VM).

**Reply by yejinchu** [reply]
get it

**Comment by liudonghua123** [comment]
After some tries, I run my pyqt5 app which is build on win11 and python 3.10 on another win7 successfully.

![image](https://user-images.githubusercontent.com/2276718/180490043-1e428c69-9d87-41a1-8f33-b3b21b18b322.png)

All I need to do is just ship the correct arch of api-ms-win-core-path-l1-1-0.dll from https://github.com/nalexandru/api-ms-win-core-path-HACK/releases/download/0.3.1/api-ms-win-core-path-blender-0.3.1.zip.

So maybe I can build a 32-bit python 3.10 based app which is compatible with more windows devices.

And during the process, I got https://github.com/pyinstaller/pyinstaller/issues/4855 too, simple install [KB3126587](https://www.catalog.update.microsoft.com/Search.aspx?q=KB3126587) to fix it.

**Reply by taurus1211** [reply]
fixed my problem!  niu bi !

**Comment by liudonghua123** [comment]
I tested with both 32-bit and 64-bit of the latest python 3.10.5 embeddable package/installer package (Because you can't install the installer package, you can install on a supported os then copy it). 

If I added the correct arch of `api-ms-win-core-path-l1-1-0.dll` from https://github.com/nalexandru/api-ms-win-core-path-HACK/releases/download/0.3.1/api-ms-win-core-path-blender-0.3.1.zip to the location of `python3.dll`, then it worked.

https://user-images.githubusercontent.com/2276718/180608663-a4cfa152-b44a-4e61-aea2-fcaccd4657ca.mp4

And If you encounter some problems like `The procedure entry point ucrtbase.terminate could not be located in the dynamic link library api-ms-win-crt-runtime-l1-1-0.dll` or `The program can’t start because api-ms-win-crt-runtime-l1-1-0.dll is missing from your computer`. You need to install `Microsoft Visual C++ Redistributable` ([32-bit](https://aka.ms/vs/17/release/vc_redist.x86.exe), [64-bit](https://aka.ms/vs/17/release/vc_redist.x64.exe)) from https://docs.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170.

**Reply by taurus1211** [reply]
niu bi plus!
